### PR TITLE
Add GafaelfawrLogoutHandler

### PR DESCRIPTION
**Pull request by athornton**

### CODE REVIEW

File: requirements/main.txt [hunk 1]
```diff
@@ -402,9 +407,9 @@ jupyterhub-idle-culler==1.2.1 \
     --hash=sha256:c84e45a51932a34bd95c08b3b6e8330fb0ee3391fa4d0f1ac0d4458a40492fd9 \
     --hash=sha256:d80822982d2590cb876d01849cee028943fdbdcc991c0f02fce45af3f40dd415
     # via -r requirements/main.in
-jupyterhub-kubespawner==4.0.0 \
-    --hash=sha256:09d6005aa722a99f0a7e3127c6029465e5603451d4c0077fcef32834dba2892a \
-    --hash=sha256:d06a6f90711c6748fc48f681994f04f1240d23117c341b1f3ac9897fd9ef3da2
+jupyterhub-kubespawner==4.1.0 \
```
Comment (rra) [inline requirements/main.txt:410]: You're upgrading jupyterhub-kubespawner here; are you sure that's okay? (Is this something we should generally pin and update explicitly or allow to float?)

File: src/nublado2/auth.py [hunk 2]
```diff
@@ -140,8 +140,11 @@ async def authenticate(
         raise NotImplementedError()
 
     def get_handlers(self, app: JupyterHub) -> List[Route]:
-        """Register the header-only login handler."""
-        return [("/gafaelfawr/login", GafaelfawrLoginHandler)]
+        """Register the header-only login and the logout handlers."""
+        return [
+            ("/gafaelfawr/login", GafaelfawrLoginHandler),
+            ("/logout", GafaelfawrLogoutHandler),
```
Comment (rra) [inline src/nublado2/auth.py:146]: Oh, good, I was hoping that would work.

File: src/nublado2/auth.py [hunk 3]
```diff
@@ -178,6 +181,23 @@ async def refresh_user(
...
+    A logout should always stop all running servers, and then redirect to the
+    RSP logout page.
+    """
+
+    @property
+    def shutdown_on_logout(self) -> bool:
+        """Unconditionally true for Gafaelfawr logout"""
+        return True
+
+    async def render_logout_page(self):
```
Comment (rra) [inline src/nublado2/auth.py:196]: Missing type annotation for the return value.

**Review by rra — COMMENTED** [review]
Looks good overall.

File: src/nublado2/auth.py (hunk 2 above)
Comment (rra) [reply src/nublado2/auth.py:146]: Just to check, did you verify that this correctly deletes the user's cookie and their auth database entry? (In other words, check that the regular `LogoutHandler` code still runs and we're not accidentally disabling it.)

Comment (athornton) [reply src/nublado2/auth.py:146]: I did: default_handle_logout (which runs before our logout_handler) shuts down the user servers and does the _backend_logout_cleanup.

File: requirements/main.txt (hunk 1 above)
Comment (athornton) [reply requirements/main.txt:410]: Now that we're just using upstream, I'm happy with allowing it to float.  They're good about matching JH versions.

File: src/nublado2/auth.py (hunk 3 above)
Comment (athornton) [reply src/nublado2/auth.py]: Really, it was a stray "return True" and it should return `None`.  But yes.  Fixed.

**Pull request merged**